microcloud/charms/src/charm.py

```diff
@@ -0,0 +1,677 @@
+            return
+
+        # Apply all the configs that changed
+        try:
+            if (
+                "snap-channel-lxd" in changed
+                or "snap-channel-microcloud" in changed
+                or "snap-channel-microceph" in changed
+                or "snap-channel-microovn" in changed
+            ):
+                logger.info(
+                    "Changes have been detected in the snap channels, updating the snaps..."
+                )
+                self.snap_install_microcloud()
+        except RuntimeError:
+            msg = "Failed to apply some configuration change(s): %s" % ", ".join(changed)
+            self.unit_blocked(msg)
+            event.defer()
+            return
+
+    def _on_cluster_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Add a new node to the existing Microcloud cluster."""
+        if (
+            self.unit.is_leader()
+            and self.get_peer_data_str(self.unit, "clustered") == "True"
+            and event.unit
+            != self.unit  # Don't add the leader to the cluster as it is already there
+        ):
+            try:
+                self.microcloud_add()
+                logger.info("New Microcloud node successfully added")
+                return
+            except RuntimeError:
+                logger.error("Failed to add a new Microcloud node")
+                return
+
+    def _on_cluster_relation_departed(self, event: RelationDepartedEvent) -> None:
+        """Update the application information regarding the number of clustered units."""
+        if self.unit.is_leader():
+            num_clustered_units = int(self.get_peer_data_str(self.app, "num_clustered_units"))
+            self.set_peer_data_str(self.app, "num_clustered_units", str(num_clustered_units - 1))
+
+    def _on_cluster_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Effectively remove this node from the existing Microcloud cluster."""
+        if self.get_peer_data_str(self.unit, "clustered") == "True":
+            try:
+                self.microcloud_remove(os.uname().nodename)
+                logger.info("Microcloud node successfully removed")
+            except RuntimeError:
+                logger.error("Failed to remove a Microcloud node, retrying later")
+
+    def config_changed(self) -> Dict:
+        """Figure out what changed."""
+        new_config = self.config
+        old_config = self._stored.config
+        apply_config = {}
+        for k, v in new_config.items():
+            if k not in old_config:
+                apply_config[k] = v
+            elif v != old_config[k]:
+                apply_config[k] = v
+
+        return apply_config
+
+    def config_is_valid(self) -> bool:
+        """Validate the config."""
+        config_changed = self.config_changed()
+        logger.info(f"Validating config: {config_changed}")
+
+        # TODO: For now, we don't have any config to validate
```

Comment:
Enabling/disabling `microovn`/`microceph` on an already deployed app should probably be rejected.